File: 1816121032928874928.py
import pandas as pd
import numpy as np
import requests
from io import BytesIO
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.dates as mdates
from datetime import datetime
from matplotlib.ticker import FuncFormatter

# Step 1: Fetch the data
url = "https://www.bdm.insee.fr/series/sdmx/data/SERIES_BDM/010768261+010768307"
response = requests.get(url)
data = pd.read_xml(BytesIO(response.content))

# Step 2: Convert to DataFrame and perform initial data manipulation
data = pd.DataFrame(data)
data['date'] = pd.to_datetime(data['TIME_PERIOD'] + "-01")
data = data[data['date'] >= "2017-05-01"]

# Step 3: Clean and transform data
data['OBS_VALUE'] = np.where((data['date'].dt.year == 2020) & (data['date'].dt.month.isin([3, 4, 5, 6])), np.nan, data['OBS_VALUE'].astype(float))
data['Naf2'] = data['TITLE_FR'].str.extract(r"(?<=- ).*(?= \(NAF)")

# Group and normalize
data = data.groupby('Naf2').apply(lambda group: group.assign(OBS_VALUE=100 * group['OBS_VALUE'] / group.loc[group['date'] == '2017-05-01', 'OBS_VALUE'].values[0]))
data.reset_index(drop=True, inplace=True)

# Step 4: Plotting
plt.figure(figsize=(1.25*6, 1.25*3.375))
sns.set(style="whitegrid")
palette = sns.color_palette("viridis", len(data['Naf2'].unique()))

# Line plot
sns.lineplot(data=data, x='date', y='OBS_VALUE', hue='Naf2', palette=palette)

# Add labels
last_values = data.loc[data.groupby('Naf2')['date'].idxmax()]
for line in last_values.itertuples():
    plt.text(line.date, line.OBS_VALUE, f"{line.OBS_VALUE/100-1:.1%}", horizontalalignment='left', size='small', color=line.Naf2)

# Add horizontal line at y=100
plt.axhline(y=100, color='grey', linestyle='--')

# Highlight period of interest
plt.axvspan(pd.to_datetime("2020-02-01"), pd.to_datetime("2020-07-01"), color=sns.color_palette("viridis", 4)[3], alpha=0.2)

# Formatting the x-axis
plt.gca().xaxis.set_major_locator(mdates.YearLocator())
plt.gca().xaxis.set_major_formatter(mdates.DateFormatter('%Y'))
plt.xticks(rotation=45)

# Formatting the y-axis
plt.gca().yaxis.set_major_formatter(FuncFormatter(lambda y, _: f"{y-100:.1%}"))

# Title and labels
plt.title("Indice de Production Industrielle vs. Mai 2017\nhors Mars-Juin 2020")
plt.xlabel("")
plt.ylabel("")

# Adjust legend
plt.legend(title='', loc='upper left')

# Save plot
plt.savefig("1816121032928874928.png", dpi=300, bbox_inches='tight', facecolor='white')
plt.savefig("1816121032928874928.pdf", dpi=300, bbox_inches='tight')

plt.show()

File: 1816121032928874928.R
# Indice de production industrielle vs. Mai 2017
# @FrancoisGeerolf, status: 1816121032928874928
# https://twitter.com/FrancoisGeerolf/status/1816121032928874928

library("tidyverse")

data <- "010768261+010768307" |>
  paste0("https://www.bdm.insee.fr/series/sdmx/data/SERIES_BDM/", a = _) |>
  rsdmx::readSDMX() |>
  as_tibble()  |>
  mutate(date = as.Date(paste0(TIME_PERIOD, "-01"))) |>
  filter(date >= as.Date("2017-05-01")) |>
  mutate(OBS_VALUE = ifelse(year(date) == 2020 & month(date) %in% c(3, 4, 5, 6), NA, as.numeric(OBS_VALUE)),
         Naf2 = str_extract(TITLE_FR, "(?<=- ).*(?= \\(NAF)")) |>
  group_by(Naf2) |>
  mutate(OBS_VALUE = 100*OBS_VALUE/OBS_VALUE[date == as.Date("2017-05-01")])

data |>
  ggplot() + ylab("Indice de Production Industrielle vs. Mai 2017\nhors Mars-Juin 2020") + xlab("") + theme_minimal() +
  geom_line(aes(x = date, y = OBS_VALUE, color = Naf2))  +
  geom_label_repel(data = data |>
                     filter(date == max(date)),
                   aes(x = date, y = OBS_VALUE, label = percent(OBS_VALUE/100-1, acc = .1), color = Naf2)) +
  geom_hline(yintercept = 100, linetype = "dashed") +
  scale_x_date(breaks = as.Date(paste0(seq(1920, 2100, 1), "-01-01")),
               labels = date_format("%Y")) +
  theme(legend.position = c(0.7, 0.9),
        legend.title = element_blank()) +
  scale_y_log10(breaks = seq(90, 110, 1),
                labels = percent(seq(90, 110, 1)/100 - 1, 1)) +
  geom_rect(data = data_frame(start = as.Date("2020-02-01"), 
                              end = as.Date("2020-07-01")), 
            aes(xmin = start, xmax = end, ymin = 0, ymax = +Inf), fill = viridis(4)[4], alpha = 0.2)

ggsave("1816121032928874928.png", height = 1.25*3.375, width = 1.25*6, bg = "white")
ggsave("1816121032928874928.pdf", height = 1.25*3.375, width = 1.25*6)